src/EditorFeatures/Core/Implementation/IntelliSense/Completion2/RoslynAsyncCompletionService.cs

```diff
@@ -0,0 +1,587 @@
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things.  Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>). 
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(snapshot, applicableToSpan))
+                {
+                    return Task.FromResult(new EditorCompletion.CompletionList(ImmutableArray<CompletionItemWithHighlight>.Empty, 0));
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var activeFilters = filters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = activeFilters.Length > 0 && activeFilters.Length < filters.Length;
+
+            // Cache PatternMatchers across this round of filtering
+            var patternMatcherMap = new Dictionary<(string pattern, CultureInfo, bool includeMatchedSpans), PatternMatcher>();
+
+            var filterResults = new List<FilterResult>();
+            foreach (var item in sortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, activeFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item, filterText, triggerReason, filterReason, patternMatcherMap))
+                {
+                    filterResults.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (triggerReason == CompletionTriggerReason.Deletion ||
+                        triggerReason == CompletionTriggerReason.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        filterResults.Add(new FilterResult(item, filterText, matchedFilterText: false));
+                    }
+                }
+            }
+
+            if (filterResults.Count == 0)
+            {
+                return Task.FromResult(HandleAllItemsFilteredOut(triggerReason, filters, activeFilters));
+            }
+
+            // If this was deletion, then we control the entire behavior of deletion
+            // ourselves.
+            if (triggerReason == CompletionTriggerReason.Deletion)
+            {
+                return HandleDeletionTrigger(triggerReason, filters, filterReason, filterText, filterResults, patternMatcherMap);
+            }
+
+            return HandleNormalFiltering(snapshot, filterText, filterResults, patternMatcherMap);
+        }
+
+        private IEnumerable<CompletionItemWithHighlight> GetHighlightedList(
```

Comment:
How did the match span highlighting work in the old system? Did we have to remember to add that information every time?